org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs

```diff
@@ -0,0 +1,718 @@
+                var client = new RestClient(url);
+                client.Timeout = -1;
+                var request = new RestRequest(Method.PATCH);
+                request.AddHeader("accept", "application/json");
+                request.AddHeader("content-type", "application/json");
+                request.AddHeader("Authorization", $"Bearer {key}");
+                request.AddParameter( "application/json", $"{{\"properties\": {{ {String.Join(",", properties.Select(p => $"\"{p.property}\": \"{p.value}\""))} }} }}", ParameterType.RequestBody);
+                IRestResponse response = client.Execute(request);
+                if ( ( int )response.StatusCode == 429 ) // To many requests
+                {
+                    if ( attempt < 3 )
+                    {
+                        Thread.Sleep(9000);
+                        UpdateHsContact(current_id, url, properties, attempt + 1);
+                    }
+                }
+                if ( response.StatusCode != HttpStatusCode.OK )
+                {
+                    throw new Exception(response.Content);
+                }
+                if ( ( int )response.StatusCode == 409 || ( int )response.StatusCode == 400 )
+                {
+                    Debug.WriteLine( string.Format("********* UpdateHsContact() HTTP STATUS CODE: {0}} | {1}", ( int )response.StatusCode, url) );
+                }
+
+            }
+            catch ( Exception e )
+            {
+                var json = $"{{\"properties\": {JsonConvert.SerializeObject(properties)} }}";
+                ExceptionLogService.LogException(new Exception($"Hubspot Sync Error{Environment.NewLine}{e}{Environment.NewLine}Current Id: {current_id}{Environment.NewLine}Exception from Request:{Environment.NewLine}{e.Message}{Environment.NewLine}Request:{Environment.NewLine}{json}{Environment.NewLine}"));
+            }
+        }
+
+        private void WriteToLog(string message)
+        {
+            string logFile = System.Web.Hosting.HostingEnvironment.MapPath("~/App_Data/Logs/HubSpotMatchLog.txt");
+            using ( System.IO.FileStream fs = new System.IO.FileStream(logFile, System.IO.FileMode.Append, System.IO.FileAccess.Write) )
+            {
+                using ( System.IO.StreamWriter sw = new System.IO.StreamWriter(fs) )
+                {
+                    sw.WriteLine(message);
+                }
+            }
+
+        }
+
+        private void GetContacts(string url)
+        {
+            request_count++;
+            var contactClient = new RestClient(url);
+            contactClient.Timeout = -1;
+            var contactRequest = new RestRequest(Method.GET);
+            contactRequest.AddHeader("Authorization", $"Bearer {key}");
+            IRestResponse contactResponse = contactClient.Execute(contactRequest);
+            var contactResults = JsonConvert.DeserializeObject<HSContactQueryResult>(contactResponse.Content);
```

Comment:
Can we add error handling to this so that if HubSpot is unresponsive, our key is no longer valid, or some other issue happens with this request the job fails gracefully instead of trying to use bad/nonexistent data?
This could be in the form of checking the status code of the response to make sure it starts with a 2 and throwing an error if it doesn't.